Announcements

billyeh opened the pull request.

radicalray commented on ap/announcements/templates/announcement_detail/table.html:
```diff
@@ -0,0 +1,28 @@
+<table class="table">
+    <tr><td>Status:</td>
+        <td>{{ announcement.get_status_display }}</td></tr>
```
Are these tabs or 4 spaces? @billyeh, Can you change all the spacing to 2-space tabs for consistency? Sublime can do that automatically for you (convert space to tabs, change tab size and then convert back to spaces) There might be faster ways to do this in bulk but this is the way I know on top of my head.

billyeh replied:
K, changed all the files in this pull request to 2 spaces :) was 4 before in some parts

radicalray commented on ap/announcements/views.py:
```diff
@@ -0,0 +1,163 @@
+
+    def get_queryset(self):
+        announcements = Announcement.objects \
+        .filter(Q(type='CLASS',
```
I don't think you need Q, you can just do ```.filter(type='CLASS', status='A', ...)```

radicalray commented on ap/aputils/trainee_utils.py:
```diff
@@ -1,13 +1,19 @@
 
 def is_trainee(user):
-    t = user.type
+    try:
```
All users should have a type, how can this throw? If user doesn't have a type, then is_trainee is being passed the wrong type of object. This can have undesired behavior. If the template does ```{% if not is_TA(user) %} <-- display trainee info (e.g. user.type) --> {% end %}``` this will cause silent errors.

billyeh replied:
AnonymousUser which you get if you're not logged in doesn't have a type. I removed the try/catch and just handle this the same way as `exams/context_processor.py` used to, just check if there's a type attr on user. I'm not sure if it's better, maybe what we should do is have a more explicit exception to catch so that we don't have the silent errors.

radicalray commented on ap/aputils/trainee_utils.py:
```diff
@@ -1,13 +1,19 @@
 
 def is_TA(user):
-    t = user.type
+    try:
```
Same goes here.

radicalray commented on ap/templates/popups/in_class_popup.html:
```diff
@@ -0,0 +1,30 @@
+            <div class="modal-content">
+                <div class="modal-header">
+                    <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
+                    <h4 class="modal-title">You are using the server in class</h4>
```
We can make this more personal: ```{% trainee.fullname %}, you are using the server in class```

radicalray commented on ap/templates/requests/request_list.html:
```diff
@@ -0,0 +1,85 @@
+
+<a href="{{ create_url }}">Create new {{ item_name }} request</a>
+
+{% regroup object_list by get_status_display as sorted %}
```
^ this is handy

billyeh replied:
Yeah, hoping we can re-use this for web requests, A/V, room reservation, etc.

radicalray commented on ap/announcements/templates/announcement_list/title.html:
```diff
@@ -0,0 +1 @@
+{{ request.get_type_display }}
```
Any reason this is a one line file? May be clearer to just have this wherever it's referenced unless you foresee the logic here getting really messy in the future?

billyeh replied:
This is the alternative view to `announcements/templates/ta_title.html` that gets inserted into a request list. I'm not sure of a better way to do this. It's kind of unfortunate because each "request" model may have a different name for the "type" field which we want to display here. E.g. right now it's `type` for announcements and `reason` for web access. I don't think this'll complicate things too much though.

radicalray replied:
See comment below

radicalray commented on ap/announcements/models.py:
```diff
@@ -0,0 +1,55 @@
+        ordering = ['-announcement_date']
+
+    ANNOUNCE_STATUS = (
+        ('A', 'Approved'),
```
I think you reference this dictionary elsewhere to translate status to text. You can do ```.get_status_display``` to show the display name and then lowercase it for your text. For future references, you can always do: ```. get_ATTR_display``` (See: https://docs.djangoproject.com/en/1.8/ref/models/instances/#django.db.models.Model.get_FOO_display)

billyeh replied:
Cool, thanks! Hmm I don't see anywhere where I reference ANNOUNCE_STATUS with a quick grep, though. I'll try and see if I remember anything. Seems pretty inconvenient if I have used this dictionary somewhere.

radicalray replied:
Added comment to point out where you can use this. I think there was another place but don't remember.

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,86 @@
+        return []
+    trainee = trainee_from_user(request.user)
+    today = datetime.date.today()
+    announcements = Announcement.objects \
```
You should try to sort queries here instead of the template. it's way faster (postgres > python)

billyeh replied:
OK. Sorry, I'm not sure where sort is used in templates that you're referring to. 

The only instance I remember is the regroup in the request list template on request status, which is a little different.

radicalray replied:
I think that's the one ^ I was thinking of but that may not be the one this is referring to. My bad. This is more of a FYI comment then.

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,86 @@
+    notifications = []
+    if is_trainee(request.user):
+        trainee = trainee_from_user(request.user)
+        notifications.extend(discipline_announcements(trainee))
```
What's preventing you from just concat'ing the lists here? ```notifications = discipline_announcments(trainee) + server_announcements(trainee) + ...``` Reads clearer.

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,86 @@
+
+def request_statuses(trainee):
+    requests = []
+    requests.extend(IndividualSlip.objects.filter(trainee=trainee, status='F'))
```
same here

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,86 @@
+def server_announcements(trainee):
+    today = datetime.date.today()
+    announcements = Announcement.objects \
+        .annotate(num_trainees=Count('trainees')) \
```
You may want to pull this complex query out and isolate it if things change in the future, it'll be easier to debug and update.

radicalray commented on ap/announcements/models.py:
```diff
@@ -0,0 +1,55 @@
+    type = models.CharField(max_length=5, choices=ANNOUNCE_TYPE, default='CLASS')
+
+    date_requested = models.DateField(auto_now_add=True)
+    trainee = models.ForeignKey(Trainee, null=True)
```
Is trainee the author of Announcement?, the naming is a little confusing. trainee and then trainees below. Consider rename for clarity?

radicalray replied:
What do you think about renaming trainee to author or creator for clarity? Yea, Leaveslips and web access requests should probably likewise have had better naming as well but we didn't catch it earlier. ;)

billyeh replied:
Hmm, I made some comments below. Maybe what we need is an interface for requests. Getters and setters for this kind of thing, rather than relying on the field name being the same, which is what we kind of do now to avoid code duplication.

billyeh commented:
Thank you for your feedback! I did my best to address your comments.

radicalray commented on ap/announcements/views.py:
```diff
@@ -0,0 +1,162 @@
+    context['create_url'] = Announcement.get_create_url()
+    context['detail_template'] = 'announcement_list/description.html'
+    if is_TA(self.request.user):
+      context['item_title_template'] = 'announcement_list/ta_title.html'
```
Is the branching logic pulled into the view for security concerns? Not exposing TA view logic in the template? If this is pushed to template as a ```{% if is_TA %} <ta_title.html> {% else %} <title.html> {% end %}```, it'll get rid of the one-line extra template files?

billyeh replied:
If we have a `Request` base class then this will go right in the template. Otherwise, the current one line in the template works for announcements but not web access because it needs `get_reason_display`, not `get_type_display`. (See below suggestion).

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,78 @@
+  notifications = []
+  if is_trainee(request.user):
+    trainee = trainee_from_user(request.user)
+    notifications += discipline_announcements(trainee)
```
Minor preference: 
```
notifications += discipline_announcements(trainee) + \
                 server_announcements(trainee) + \
                 ...
```                          
May be clearer. I hope the backslash works.

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -12,75 +12,67 @@
+    .filter(Q(type='SERVE',
+      status='A',
+      announcement_date__lte=today,
+      is_popup=True) \
```
You can make is_popup a boolean argument into ```.announcements_for_today``` to pull this logic out too.

billyeh replied:
Only thing is I didn't plan on using `announcement_end_date` in popup announcements, but I remove that field for popups anyways. So this is great.

radicalray commented on ap/announcements/views.py:
```diff
@@ -16,148 +15,148 @@
+  name = announcement.trainee
+  message = "%s's %s web request was " % (name, announcement.get_type_display())
+  if status == 'A':
+    message += 'approved.'
```
This could use the ```.get_status_display()``` I mentioned earlier. ```announcement.get_status_display().lowercase()```

billyeh replied:
Yeah that's really good, actually, we can get rid of all those if's. Actually, this code was copied from `web_access/views.py`. I think a `Request` base class for web access, announcements, and all the other requests we plan to support would be a good thing.

radicalray commented on ap/aputils/trainee_utils.py:
```diff
@@ -1,25 +1,19 @@
-        return True
-    return False
+  t = user.type
+  if t == 'T':
```
You can just do ```return t == 'T'```

radicalray commented on ap/aputils/trainee_utils.py:
```diff
@@ -1,25 +1,19 @@
-        return True
-    return False
+  t = user.type
+  if t in ('R', 'C', 'S'):
```
Same here like comment below. Much better. 👍

radicalray commented on ap/templates/popups/in_class_popup.html:
```diff
@@ -1,30 +1,30 @@
+      <div class="modal-content">
+        <div class="modal-header">
+          <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
+          <h4 class="modal-title">{{ trainee.fullname }}, you are using the server in class</h4>
```
👍

billyeh commented:
OK. I think I addressed all your comments this time. Thank you for putting in work on a weekend! :) I went with all the suggestions you had. Maybe when we get around to implementing all these kind of requests we can see if a superclass is a good way to reduce duplicated code, since the workflow is really similar for a lot of these things.

billyeh commented:
@jontey I think we're ready :)

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,70 @@
+  if is_trainee(request.user):
+    trainee = trainee_from_user(request.user)
+    notifications += discipline_announcements(trainee) \
+        + server_announcements(trainee) \
```
Please line up the indentation to discipline?

radicalray commented on ap/announcements/notifications.py:
```diff
@@ -0,0 +1,70 @@
+
+def request_statuses(trainee):
+  requests = []
+  requests += IndividualSlip.objects.filter(trainee=trainee, status='F')
```
multi-line + here as well.

billyeh commented:
All right. I have discovered the `chain` function, awesome haha. Apparently more efficient because it's `itertools` too. http://stackoverflow.com/questions/431628/how-to-combine-2-or-more-querysets-in-a-django-view

radicalray commented:
Super! Could you add this to the wiki? Maybe under django queryset or something to help other Devs in the future. @billyeh

billyeh commented:
Sure. Done!
